corehq/apps/es/management/commands/elastic_sync_multiplexed.py

```diff
@@ -190,6 +190,38 @@ def _get_current_and_older_index_name(cls, cname):
         older_index = getattr(es_consts, f"HQ_{cname.upper()}_INDEX_NAME")
         return (current_index, older_index)
 
+    def set_checkpoints_for_new_index(self, cname):
+        """
+        Takes in an index cname and create new checkpoint for all the pillows that use the older index name
+        for that cname.
+        Can only be performed when indexes are still multiplexed and not swapped.
+        When we swap the indexes, the primary index changes which updates the checkpoint names.
+        We should stop the pillows and copy the checkpoint to the new checkpoint ids, swap the indexes
+        and then start the pillows.
```

Comment:
This sounds like a very precise set of steps that must be followed exactly or risk a bad outcome. Is there any way we can automate it or put guard rails around it to guide operators through the necessary steps in a way that is not error-prone?

Does this only involve stopping pillows, or might it require true down-time? Like is there anything other than pillows that can interact with Kafka and update a checkpoint?

Reply:
> This sounds like a very precise set of steps that must be followed exactly or risk a bad outcome.

I think the worst that can happen here is being your pillows pushed to an earlier date and reprocess that data that has already been processed. So it will lead to a pillow lag and not loosing the data.

Even if the command is run multiple times then also we will get into the same situation where pillows will show a bit of lag. 

The guard has been placed in order to ensure that it is being run in a very specific condition when the Indexes are multiplexed but not swapped.

> Does this only involve stopping pillows, or might it require true down-time? Like is there anything other than pillows that can interact with Kafka and update a checkpoint?

I don't think there is anything else that would update the checkpoint for these pillows. The process can even be run without stopping pillows. It will just result in reprocessing some of the docs. The stopping of pillows ensures that we start from the same checkpoints that earlier checkpoint id was pointing to.